clients/drcachesim/tools/func_view.h

```diff
@@ -0,0 +1,117 @@
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#ifndef _FUNC_VIEW_H_
+#define _FUNC_VIEW_H_ 1
+
+#include <mutex>
+#include <set>
+#include <set>
+#include <string>
+#include <unordered_map>
+
+#include "analysis_tool.h"
+#include "raw2trace.h"
+#include "raw2trace_directory.h"
+
+class func_view_t : public analysis_tool_t {
+public:
+    func_view_t(const std::string &funclist_file_path, bool full_trace,
+                unsigned int verbose = 0);
+    virtual ~func_view_t();
+    std::string
+    initialize() override;
+    bool
+    process_memref(const memref_t &memref) override;
+    bool
+    print_results() override;
+    bool
+    parallel_shard_supported() override;
+    void *
+    parallel_shard_init(int shard_index, void *worker_data) override;
+    bool
+    parallel_shard_exit(void *shard_data) override;
+    bool
+    parallel_shard_memref(void *shard_data, const memref_t &memref) override;
+    std::string
+    parallel_shard_error(void *shard_data) override;
+
+protected:
+    struct func_stats_t {
+        func_stats_t &
+        operator+=(const func_stats_t &rhs)
+        {
+            num_calls += rhs.num_calls;
+            num_returns += rhs.num_returns;
+            return *this;
+        }
+        int_least64_t num_calls = 0;
+        int_least64_t num_returns = 0;
+        // TODO i#4083: Record the arg and retval distributions.
+    };
+    struct shard_data_t {
+        memref_tid_t tid = 0;
+        std::unordered_map<int, func_stats_t> func_map;
+        std::string error;
+        int last_func_id = -1;
+        int nesting_level = 0;
+        bool prev_was_arg = false;
+        addr_t prev_pc = 0;
+        app_pc last_trace_module_start = nullptr;
+        size_t last_trace_module_size = 0;
+        std::string last_trace_module_name;
+    };
+
+    static bool
+    cmp_func_stats(const std::pair<int, func_stats_t> &l,
+                   const std::pair<int, func_stats_t> &r);
+    void
+    process_memref_for_markers(void *shard_data, const memref_t &memref);
+    std::unordered_map<int, func_stats_t>
+    compute_totals();
+
+    static const std::string TOOL_NAME;
+    bool knob_full_trace_;
+    unsigned int knob_verbose_;
+
+    std::unordered_map<int, std::set<std::string>> id2name_;
+
+    std::string funclist_file_path_;
+
+    // The keys here are shard index for parallel, tid for serial.
+    std::unordered_map<memref_tid_t, shard_data_t *> shard_map_;
+    // This mutex is only needed in parallel_shard_init.  In all other accesses to
```

Comment:
Your call, might still be worth grabbing in the other cases so it's easier to make tsan happy in the future.

Reply:
This is how the other tools use the lock as well and tsan does not complain.  I just ran this new one (internally w/ adapters) under tsan as well and it does not report any problems. (Confirmed that removing the lock from parallel_shared_init does result in data races reported.)  I am inclined to leave it to match the others and avoid extra locks.